internal/tpm/csr.go

```diff
@@ -0,0 +1,503 @@
+		return nil, err
+	}
+
+	// Build content structure with sizes
+	content := IDevIDContent{
+		StructVer:                 [4]byte{0x00, 0x00, 0x01, 0x00},         // Version 1.0
+		HashAlgoId:                [4]byte{0x00, 0x00, 0x00, TCGAlgSHA256}, // SHA256
+		HashSz:                    [4]byte{0x00, 0x00, 0x00, 32},           // SHA256 hash size
+		ProdModelSz:               uint32ToBytes(safeIntToUint32(len(payload.ProdModel))),
+		ProdSerialSz:              uint32ToBytes(safeIntToUint32(len(payload.ProdSerial))),
+		ProdCaDataSz:              uint32ToBytes(safeIntToUint32(len(payload.ProdCaData))),
+		BootEvntLogSz:             uint32ToBytes(safeIntToUint32(len(payload.BootEvntLog))),
+		EkCertSz:                  uint32ToBytes(safeIntToUint32(len(payload.EkCert))),
+		AttestPubSz:               uint32ToBytes(safeIntToUint32(len(payload.AttestPub))),
+		AtCreateTktSz:             uint32ToBytes(safeIntToUint32(len(payload.AtCreateTkt))),
+		AtCertifyInfoSz:           uint32ToBytes(safeIntToUint32(len(payload.AtCertifyInfo))),
+		AtCertifyInfoSignatureSz:  uint32ToBytes(safeIntToUint32(len(payload.AtCertifyInfoSignature))),
+		SigningPubSz:              uint32ToBytes(safeIntToUint32(len(payload.SigningPub))),
+		SgnCertifyInfoSz:          uint32ToBytes(safeIntToUint32(len(payload.SgnCertifyInfo))),
+		SgnCertifyInfoSignatureSz: uint32ToBytes(safeIntToUint32(len(payload.SgnCertifyInfoSignature))),
+		PadSz:                     uint32ToBytes(safeIntToUint32(len(payload.Pad))),
+	}
+
+	// Serialize content and payload for hashing
+	contentBytes, err := serializeTCGContent(content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize TCG content: %w", err)
+	}
+
+	payloadBytes, err := serializeTCGPayload(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize TCG payload: %w", err)
+	}
+
+	// Create data to be hashed (content + payload)
+	dataToHash := append(contentBytes, payloadBytes...)
+
+	// Hash the data using SHA256 since TPM Sign expects a digest, not raw data
+	hash := crypto.SHA256.New()
+	hash.Write(dataToHash)
+	digest := hash.Sum(nil)
+
+	// Sign the digest
+	signature, err := signer.Sign(nil, digest, crypto.SHA256)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign TCG-CSR data: %w", err)
+	}
+
+	// Check combined content and payload size
+	combinedSize := len(contentBytes) + len(payloadBytes)
+	if combinedSize > maxTotalSize {
+		return nil, fmt.Errorf("combined content and payload size exceeds maximum: %d bytes (max %d)", combinedSize, maxTotalSize)
+	}
+
+	// Build final TCG-CSR-IDEVID structure
+	tcgCSR := TCGCSRIDevID{
+		StructVer: [4]byte{0x01, 0x00, 0x01, 0x00}, // Version 1.0
+		Contents:  uint32ToBytes(safeIntToUint32(combinedSize)),
+		SigSz:     uint32ToBytes(safeIntToUint32(len(signature))),
+		Signature: signature,
+	}
+
+	// Serialize the complete structure
+	result := &bytes.Buffer{}
+
+	// Write header
+	result.Write(tcgCSR.StructVer[:])
+	result.Write(tcgCSR.Contents[:])
```

Comment:
If it doesn't violate what the spec has named these fields can we call this ContentSz? Initially it looked like the header was writing the actual contents.